### PR TITLE
[AutoDiff] Fix SR-12641: Handle address-only types in derivative fn types

### PR DESCRIPTION
**Pull request by marcrasi**
* The update in `SILFunctionType.cpp` fixes SR-12641 by making address-only parameters/results in differentials have indirect convention.
* I updated the crasher test to use a resilient struct defined in the test, instead of `Tracked<Float>`, so that the test does not need to depend on `DifferentiationUnittest`.
* The update in `VJPEmitter.cpp` fixes a similar issue with pullbacks that I discovered while investigating.
* I added code that exposes this new issue to the SR-12641 crasher test.

### CODE REVIEW

File: lib/SIL/IR/SILFunctionType.cpp [hunk 1]
```diff
@@ -303,7 +303,51 @@ getSemanticResults(SILFunctionType *functionType, IndexSubset *parameterIndices,
...
+  auto getTangentParameterConvention =
+      [&](CanType tanType,
+          ParameterConvention origParamConv) -> ParameterConvention {
+    tanType =
+        tanType->getCanonicalType(originalFnTy->getSubstGenericSignature());
+    AbstractionPattern pattern(originalFnTy->getSubstGenericSignature(),
+                               tanType);
+    auto &tl =
+        TC.getTypeLowering(pattern, tanType, TypeExpansionContext::minimal());
+    if (tl.isAddressOnly() && !isIndirectFormalParameter(origParamConv)) {
```
Comment (dan-zheng) [inline lib/SIL/IR/SILFunctionType.cpp:317]: An explanatory comment on top of this `if` would be nice to help readers: something about handling address-only `TangentVector` types. It took me a minute to think through why this `if` is necessary.

Same regarding the `if` below in `getTangentResultConvention`.

**Review by dan-zheng — APPROVED** [review]
Nice! This fixes some original vs tangent convention mismatch issues - do you think any holes remain in the codebase?

File: lib/SIL/IR/SILFunctionType.cpp (hunk 1 above)
Comment (marcrasi) [reply lib/SIL/IR/SILFunctionType.cpp:321]: Done

**Comment by marcrasi** [comment]
> do you think any holes remain in the codebase?

I didn't do any exhaustive thinking while writing this PR, I just stumbled on these two cases by accident. So it's possible that there are some more holes somewhere.

There is this inelegant diagnostic that simply rejects a lot of these cases: https://github.com/apple/swift/blob/a021e6ca020e667ce4bc8ee174e2de1cc0d9be73/include/swift/AST/DiagnosticsSIL.def#L500

**Comment by marcrasi** [comment]
@swift-ci please test

**Comment by dan-zheng** [comment]
Merging now to unblock progress.

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Pull request merged**